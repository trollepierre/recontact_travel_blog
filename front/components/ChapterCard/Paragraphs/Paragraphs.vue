<template>
  <footer class="chapter__footer">
    <ul
      v-for="paragraph in paragraphs"
      :key="paragraph.text"
      class="paragraph-container">
      <li
        v-if="paragraph"
        class="paragraph">
        <iframe
          v-if="paragraph.iframeSrc"
          :width="dimensions.width"
          :height="dimensions.height"
          :src="paragraph.iframeSrc"
          :title="title"
          class="youtube-iframe"
          allow="accelerometer; encrypted-media; gyroscope;"
          allowfullscreen/>
        <p
          v-else-if="paragraph.link"
          class="chapter__footer_text">
          <a
            :href="paragraph.link"
            rel="noreferrer"
            target="_blank">
            {{ paragraph.link }}
          </a>
        </p>
        <h3
          v-else
          class="chapter__footer_text">
          {{ paragraph.text }}
        </h3>
      </li>
    </ul>
  </footer>
</template>
<script>
  import { iframeDimensions } from '../../../services'
  import {
    generateCleanUrlLink,
    generateIframeLink,
    urlTester,
    youtubeEmbedUrlTester,
  } from './paragraph-link-utils'

  export default {
    name: 'Paragraphs',
    props: {
      chapterText: { type: Array, default: () => [] },
      title: { type: String, default: () => '' },
    },
    data: () => ({ dimensions: iframeDimensions() }),
    computed: {
      paragraphs() {
        return this.chapterText
          .filter(paragraph => !!paragraph)
          .map(this.enhanceParagraph)
      },
    },
    methods: {
      enhanceParagraph(paragraph) {
        return {
          iframeSrc: youtubeEmbedUrlTester(paragraph) ? generateIframeLink(paragraph) : undefined,
          link: urlTester(paragraph) ? generateCleanUrlLink(paragraph) : undefined,
          text: paragraph,
        }
      },
    },
  }
</script>

<style scoped>
.chapter__footer {
  text-align: center;
  padding: 0 15px;
  border-top: 1px solid #E6E6E6;
}

.chapter__footer button {
  text-transform: uppercase;
  color: #D14800;
  background: #FFFFFF;
  border: 1px solid #D14800;
  cursor: pointer;
  padding: 15px 30px;
  border-radius: 4px;
  width: 100%;
  margin-bottom: 10px;
  font-weight: 700;
}

.chapter__footer button:hover {
  background: #D14800;
  color: #FFFFFF;
}

.chapter__footer button:disabled,
.chapter__footer button:active {
  background: #BDBDBD;
  border-color: #616161;
  color: #FAFAFA;
  cursor: auto;
}

.youtube-iframe {
  margin: 30px 0;
  border: none;
}

.chapter__footer_text {
  font-size: 18px;
  word-spacing: 1px;
}

.paragraph {
  list-style-type: none;
}

.paragraph-container {
  padding: 0;
}

@media only screen and (min-width: 640px) {
  .chapter__footer_text {
    font-weight: lighter;
  }
}
</style>
