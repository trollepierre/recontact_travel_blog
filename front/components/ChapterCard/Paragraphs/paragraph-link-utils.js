const urlTester = link => /^(https?:\/\/)?([\da-z.-]+)\.([a-z.]{2,6})([/\w .-]*)*\/?/.test(link)

const youtubeEmbedUrlTester = link => urlTester(link) && (/youtu.be/.test(link) || /youtube.com/.test(link)) && /embed/.test(link)
const youtubePlaylistEmbedUrlTester = link => /videoseries\?list=/.test(link)
const generateIframeLink = link => youtubePlaylistEmbedUrlTester(link) ? link : `${link}?rel=0&modestbranding=1`
const generateCleanUrlLink = link => link.includes('https://') ? link : `https://${link}`

export {
  youtubeEmbedUrlTester,
  youtubePlaylistEmbedUrlTester,
  generateCleanUrlLink,
  generateIframeLink,
  urlTester,
}
