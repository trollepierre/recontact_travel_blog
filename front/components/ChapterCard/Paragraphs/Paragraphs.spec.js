import VueI18n from 'vue-i18n'
import Paragraphs from './Paragraphs.vue'

describe('Component | Paragraphs.vue', () => {
  let wrapper
  let propsData
  let localVue

  beforeEach(() => {
    console.warn = jest.fn()
    propsData = { title: 'title', chapterText: ['voilÃ ', ''] }
    localVue = createLocalVue()
    localVue.use(VueI18n)
  })

  describe('template', () => {
    it('should match snapshot with text', () => {
      wrapper = shallowMount(Paragraphs, { propsData })
      expect(wrapper).toMatchSnapshot()
    })

    it('should match snapshot with link', () => {
    propsData = { title: 'title', chapterText: ['https://www.rec.me'] }
      wrapper = shallowMount(Paragraphs, { propsData })
      expect(wrapper).toMatchSnapshot()
    })

    it('should match snapshot with video on click on loading button', async () => {
    propsData = { title: 'title', chapterText: ['https://www.youtu.be/embed/123'] }
      wrapper = shallowMount(Paragraphs, { propsData, localVue })
      await wrapper.find('button').trigger('click')
      expect(wrapper).toMatchSnapshot()
    })
  })
})

