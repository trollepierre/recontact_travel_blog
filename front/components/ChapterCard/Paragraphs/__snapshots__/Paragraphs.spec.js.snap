// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | Paragraphs.vue template should match snapshot with link 1`] = `
<footer class="chapter__footer">
  <ul class="paragraph-container">
    <li class="paragraph">
      <p class="chapter__footer_text"><a href="https://www.rec.me" rel="noreferrer" target="_blank">
          https://www.rec.me
        </a></p>
    </li>
  </ul>
</footer>
`;

exports[`Component | Paragraphs.vue template should match snapshot with text 1`] = `
<footer class="chapter__footer">
  <ul class="paragraph-container">
    <li class="paragraph">
      <h3 class="chapter__footer_text">
        voil√†
      </h3>
    </li>
  </ul>
</footer>
`;

exports[`Component | Paragraphs.vue template should match snapshot with video on click on loading button 1`] = `
<footer class="chapter__footer">
  <ul class="paragraph-container">
    <li class="paragraph"><iframe width="512" height="288" src="https://www.youtu.be/embed/123?rel=0&amp;modestbranding=1" title="title" allow="accelerometer; encrypted-media; gyroscope;" allowfullscreen="allowfullscreen" class="youtube-iframe"></iframe></li>
  </ul>
</footer>
`;
